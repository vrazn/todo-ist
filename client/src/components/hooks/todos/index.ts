import useSWR from 'swr';
import useSWRMutation from 'swr/mutation';

import type {
  ITodo,
  ITodoCreationAttributes,
  ITodoDeleteAttributes,
  ITodoUpdateAttributes,
} from 'todo-list';

const headers = {
  'content-type': 'application/json',
};

const GET = async (input: string) => {
  const response = await fetch(input, { headers });
  const json = await response.json();

  if (!response.ok) {
    throw json;
  }
  return json;
};

const POST = async (
  input: string,
  { arg }: { arg: ITodoCreationAttributes },
) => {
  const response = await fetch(input, {
    method: 'POST',
    body: JSON.stringify(arg),
    headers,
  });
  const json = await response.json();

  if (!response.ok) {
    throw json;
  }
  return json;
};

const PATCH = async (
  input: string,
  { arg }: { arg: ITodoUpdateAttributes },
) => {
  const { id, ...body } = arg;

  const response = await fetch(`${input}/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(body),
    headers,
  });
  const json = await response.json();

  if (!response.ok) {
    throw json;
  }
  return json;
};

const DELETE = async (
  input: string,
  { arg }: { arg: ITodoDeleteAttributes },
) => {
  const { id } = arg;

  const response = await fetch(`${input}/${id}`, { method: 'DELETE', headers });
  const json = await response.json();

  if (!response.ok) {
    throw json;
  }
  return json;
};

export const useGetTodos = () => {
  const { data, isLoading, error } = useSWR<ITodo[]>(
    `${process.env.NEXT_PUBLIC_BACKEND_URL}/todos`,
    GET,
    { refreshInterval: 0 },
  );

  return {
    todos: data,
    isLoading,
    error,
  };
};

export const useCreateTodo = () => {
  const { data, trigger, isMutating, error } = useSWRMutation<ITodo>(
    `${process.env.NEXT_PUBLIC_BACKEND_URL}/todos`,
    POST,
  );

  return {
    todo: data,
    trigger,
    isMutating,
    error,
  };
};

export const usePatchTodo = () => {
  const { data, trigger, isMutating, error } = useSWRMutation<ITodo>(
    `${process.env.NEXT_PUBLIC_BACKEND_URL}/todos`,
    PATCH,
  );

  return {
    todo: data,
    trigger,
    isMutating,
    error,
  };
};

export const useDeleteTodo = () => {
  const { data, trigger, isMutating, error } = useSWRMutation<ITodo>(
    `${process.env.NEXT_PUBLIC_BACKEND_URL}/todos`,
    DELETE,
  );

  return {
    todo: data,
    trigger,
    isMutating,
    error,
  };
};
