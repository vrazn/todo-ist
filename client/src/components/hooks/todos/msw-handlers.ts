import { rest } from 'msw';
import { ITodo } from 'todo-list';

export const todos = [
  {
    id: 1,
    title: 'Title 1',
    description: 'Description 1',
    isImportant: true,
    isDone: false,
    dueDate: '2023-01-01',
  },
  {
    id: 2,
    title: 'Title 2',
    description: null,
    isImportant: false,
    isDone: false,
    dueDate: null,
  },
  {
    id: 3,
    title: 'Title 3',
    description: 'Description 3',
    isImportant: true,
    isDone: true,
    dueDate: '2023-01-03',
  },
];

export const todo = todos[0];

export const emptyTodos: ITodo[] = [];

export const error = { message: 'Test error' };

export const getEmpty = rest.get('/todos', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(200), ctx.json(emptyTodos));
});

export const getSuccessStub: ITodo[] = [];

export const getSuccess = rest.get('/todos', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(200), ctx.json(todos));
});

export const getError = rest.get('/todos', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(500), ctx.json(error));
});

export const postSuccess = rest.post('/todos', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(201), ctx.json(todo));
});

export const postError = rest.post('/todos', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(500), ctx.json(error));
});

export const patchSuccess = rest.patch('/todos/:id', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(200), ctx.json(todo));
});

export const patchError = rest.patch('/todos/:id', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(500), ctx.json(error));
});

export const deleteSuccess = rest.delete('/todos/:id', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(200), ctx.json(todo));
});

export const deleteError = rest.delete('/todos/:id', (_req, res, ctx) => {
  return res(ctx.delay(100), ctx.status(500), ctx.json(error));
});
