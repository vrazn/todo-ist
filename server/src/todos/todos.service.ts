import { Injectable, NotFoundException } from '@nestjs/common';
import { snakeCase } from 'change-case';

import { Todo } from './entities/todo.entity';
import { CreateTodoDto } from './dto/create-todo.dto';
import { UpdateTodoDto } from './dto/update-todo.dto';
import { DatabaseTable } from '../database/database.decorator';
import { DatabaseService } from '../database/database.service';

@Injectable()
export class TodosService {
  constructor(
    @DatabaseTable('todos')
    private readonly db: DatabaseService<Todo>,
  ) {}

  async create(createTodoDto: CreateTodoDto): Promise<Todo> {
    const { keys, values, indexes } = Object.keys(createTodoDto).reduce(
      (acc, key, index) => {
        acc.keys.push(snakeCase(key));
        acc.values.push(createTodoDto[key as keyof CreateTodoDto]);
        acc.indexes.push(`$${index + 1}`);
        return acc;
      },
      { keys: [], values: [], indexes: [] } as {
        keys: string[];
        values: string[];
        indexes: string[];
      },
    );

    const [result] = await this.db.insert({
      query: keys.join(', '),
      where: indexes.join(', '),
      variables: values,
    });
    return result;
  }

  async findAll(): Promise<Todo[]> {
    return this.db.select({ query: '*', variables: [] });
  }

  async findOne(id: number): Promise<Todo> {
    const [result] = await this.db.select({
      query: '*',
      where: 'id = $1',
      variables: [id],
    });

    if (!result) throw new NotFoundException();
    return result;
  }

  async update(id: number, updateTodoDto: UpdateTodoDto): Promise<Todo> {
    const keys = Object.keys(updateTodoDto);
    const values = Object.values(updateTodoDto);

    const query = keys
      .map((key, index) => `${snakeCase(key)} = $${index + 1}`)
      .join(', ');

    const [result] = await this.db.update({
      query,
      where: `id = $${keys.length + 1}`,
      variables: [...values, id],
    });

    return result;
  }

  async remove(id: number): Promise<Todo> {
    const [result] = await this.db.delete({
      where: 'id = $1',
      variables: [id],
    });

    return result;
  }
}
