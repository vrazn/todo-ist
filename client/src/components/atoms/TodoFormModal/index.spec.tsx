import { fireEvent, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { setupServer } from 'msw/node';
import { SWRConfig } from 'swr';

import * as handlers from '@/components/hooks/todos/msw-handlers';
import { TodoFormModal } from './index';

const server = setupServer();

const Wrapper = ({ children }: React.PropsWithChildren) => (
  <SWRConfig value={{ dedupingInterval: 0, provider: () => new Map() }}>
    {children}
  </SWRConfig>
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

const onCancel = jest.fn().mockImplementation(() => {});

describe('TodoFormModal', () => {
  it('renders nothing when open={false}', () => {
    const { queryByRole } = render(
      <TodoFormModal open={false} onCancel={onCancel} />,
      { wrapper: Wrapper },
    );

    expect(queryByRole('dialog')).toMatchSnapshot();
  });

  it('renders without change when open={true}', () => {
    const { queryByRole } = render(
      <TodoFormModal open={true} onCancel={onCancel} />,
      { wrapper: Wrapper },
    );
    expect(queryByRole('dialog')).toMatchSnapshot();
  });

  it('handles button clicks and triggers validations', async () => {
    const { getByText, getByLabelText } = render(
      <TodoFormModal open={true} onCancel={onCancel} />,
      { wrapper: Wrapper },
    );

    const button = getByText('Save');

    await userEvent.click(button);

    await waitFor(() => {
      expect(getByText('This field is required')).toBeInTheDocument();
    });

    const input = getByLabelText('Title');

    fireEvent.change(input, {
      target: {
        value: 'Something very long here so that it exceeds 20 symbols',
      },
    });

    await waitFor(() => {
      expect(
        getByText('This field cannot be longer than 20 characters'),
      ).toBeInTheDocument();
    });

    fireEvent.change(input, {
      target: {
        value: '',
      },
    });

    await waitFor(() => {
      expect(getByText('This field is required')).toBeInTheDocument();
    });

    const closeButton = getByLabelText('Close');
    expect(closeButton).toBeInTheDocument();
    await userEvent.click(closeButton);
    expect(onCancel).toBeCalled();
  });

  describe('form submission (success)', () => {
    beforeEach(() => {
      server.use(handlers.postSuccess, handlers.patchSuccess);
    });

    it('handles create', async () => {
      const { getByText, getByLabelText } = render(
        <TodoFormModal
          initialState={undefined}
          open={true}
          onCancel={onCancel}
        />,
        { wrapper: Wrapper },
      );

      const button = getByText('Save');
      const input = getByLabelText('Title');

      fireEvent.change(input, {
        target: {
          value: 'Test Title',
        },
      });

      await userEvent.click(button);

      expect(onCancel).toBeCalled();
    });

    it('handles patch', async () => {
      const { getByText } = render(
        <TodoFormModal
          initialState={handlers.todo}
          open={true}
          onCancel={onCancel}
        />,
        { wrapper: Wrapper },
      );

      const button = getByText('Save');
      await userEvent.click(button);

      expect(onCancel).toBeCalled();
    });
  });

  describe('form submission (error)', () => {
    beforeEach(() => {
      server.use(handlers.postError, handlers.patchError);
    });

    it('handles create', async () => {
      const { getByText, getByLabelText } = render(
        <TodoFormModal
          initialState={undefined}
          open={true}
          onCancel={onCancel}
        />,
        { wrapper: Wrapper },
      );

      const button = getByText('Save');
      const input = getByLabelText('Title');

      fireEvent.change(input, {
        target: {
          value: 'Test Title',
        },
      });

      await userEvent.click(button);
    });

    it('handles patch', async () => {
      const { getByText } = render(
        <TodoFormModal
          initialState={handlers.todo}
          open={true}
          onCancel={onCancel}
        />,
        { wrapper: Wrapper },
      );

      const button = getByText('Save');
      await userEvent.click(button);
    });
  });
});
