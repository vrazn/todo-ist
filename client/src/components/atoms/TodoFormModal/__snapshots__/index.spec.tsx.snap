// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoFormModal renders nothing when open={false} 1`] = `null`;

exports[`TodoFormModal renders without change when open={true} 1`] = `
<div
  aria-modal="true"
  class="ant-modal css-dev-only-do-not-override-ixblex"
  role="dialog"
  style="width: 520px;"
>
  <div
    aria-hidden="true"
    style="width: 0px; height: 0px; overflow: hidden; outline: none;"
    tabindex="0"
  />
  <div
    class="ant-modal-content"
  >
    <button
      aria-label="Close"
      class="ant-modal-close"
      type="button"
    >
      <span
        class="ant-modal-close-x"
      >
        <span
          aria-label="close"
          class="anticon anticon-close ant-modal-close-icon"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
            />
          </svg>
        </span>
      </span>
    </button>
    <div
      class="ant-modal-body"
    >
      <form
        class="ant-form ant-form-vertical css-dev-only-do-not-override-ixblex"
      >
        <div
          class="ant-form-item css-dev-only-do-not-override-ixblex"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-ixblex"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-ixblex"
            >
              <label
                class="ant-form-item-required"
                for="title"
                title="Title"
              >
                Title
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-ixblex"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    aria-required="true"
                    class="ant-input css-dev-only-do-not-override-ixblex"
                    id="title"
                    placeholder="What needs to be done?"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item css-dev-only-do-not-override-ixblex"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-ixblex"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-ixblex"
            >
              <label
                class=""
                for="description"
                title="Description (optional)"
              >
                Description (optional)
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-ixblex"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <textarea
                    class="ant-input css-dev-only-do-not-override-ixblex"
                    id="description"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item css-dev-only-do-not-override-ixblex"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-ixblex"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-ixblex"
            >
              <label
                class=""
                for="dueDate"
                title="Due Date (optional)"
              >
                Due Date (optional)
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-ixblex"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <div
                    class="ant-picker css-dev-only-do-not-override-ixblex"
                  >
                    <div
                      class="ant-picker-input"
                    >
                      <input
                        autocomplete="off"
                        id="dueDate"
                        placeholder="Select date"
                        readonly=""
                        size="12"
                        title=""
                        value=""
                      />
                      <span
                        class="ant-picker-suffix"
                      >
                        <span
                          aria-label="calendar"
                          class="anticon anticon-calendar"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="calendar"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item css-dev-only-do-not-override-ixblex"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-ixblex"
          >
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-ixblex"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <label
                    class="ant-checkbox-wrapper ant-checkbox-wrapper-in-form-item css-dev-only-do-not-override-ixblex"
                  >
                    <span
                      class="ant-checkbox css-dev-only-do-not-override-ixblex"
                    >
                      <input
                        class="ant-checkbox-input"
                        id="isImportant"
                        type="checkbox"
                        value=""
                      />
                      <span
                        class="ant-checkbox-inner"
                      />
                    </span>
                    <span>
                      Mark as important
                    </span>
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item css-dev-only-do-not-override-ixblex"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-ixblex"
          >
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-ixblex"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <label
                    class="ant-checkbox-wrapper ant-checkbox-wrapper-in-form-item css-dev-only-do-not-override-ixblex"
                  >
                    <span
                      class="ant-checkbox css-dev-only-do-not-override-ixblex"
                    >
                      <input
                        class="ant-checkbox-input"
                        id="isDone"
                        type="checkbox"
                        value=""
                      />
                      <span
                        class="ant-checkbox-inner"
                      />
                    </span>
                    <span>
                      Mark as complete
                    </span>
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="ant-modal-footer"
    >
      <button
        class="ant-btn css-dev-only-do-not-override-ixblex ant-btn-primary ant-btn-block"
        type="button"
      >
        <span
          aria-label="check-circle"
          class="anticon anticon-check-circle"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="check-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0051.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
            />
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
          </svg>
        </span>
        <span>
          Save
        </span>
      </button>
    </div>
  </div>
  <div
    aria-hidden="true"
    style="width: 0px; height: 0px; overflow: hidden; outline: none;"
    tabindex="0"
  />
</div>
`;
